### PR TITLE
Add client integration tests for process instance modification

### LINKED ISSUE #10010: Write end-to-end (QA) tests for modification using the Java client
## Description

We should write integration tests for the process instance modification to ensure that feature is working from end to end. In the integration tests, we use the Java client to send the command and verify the result. Similar to [existing QA tests](https://github.com/camunda/zeebe/blob/main/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceTest.java).

### PR DESCRIPTION
**Pull request by korthout**
## Description

This adds end-to-end integration tests for process instance modification, which showcases that the integration between the java client and the engine works as expected. 

## Related issues

closes #10010

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java [hunk 1]
```diff
@@ -120,6 +123,8 @@ public ProcessingError tryHandleError(
       // This exception is only thrown for ProcessInstanceCreationRecord with start instructions
       rejectionWriter.appendRejection(
           typedCommand, RejectionType.INVALID_ARGUMENT, exception.getMessage());
+      responseWriter.writeRejectionOnCommand(
+          typedCommand, RejectionType.INVALID_ARGUMENT, exception.getMessage());
```
Comment (saig0) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:127]: :+1: Nice catch.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceTest.java [hunk 2]
```diff
@@ -218,4 +237,41 @@ public void shouldRejectCreateBpmnProcessByNonExistingKey() {
...
+        CLIENT_RULE
+            .getClient()
+            .newCreateInstanceCommand()
+            .processDefinitionKey(thirdProcessDefinitionKey)
+            .startBeforeElement("end1")
+            .startBeforeElement("end2")
+            .send()
+            .join();
+
+    assertThat(instance.getProcessInstanceKey()).isPositive();
```
Comment (saig0) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceTest.java:254]: :wrench: We could verify that the start instructions are really applied by using the `RecordingExporter`. For example, by checking that only the end events are activated. 

Currently, the test case would pass if the start instructions are ignored (in the client, the gateway, or the broker).

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/ModifyProcessInstanceTest.java [hunk 3]
```diff
@@ -30,14 +34,65 @@ public class ModifyProcessInstanceTest {
...
+            .activateElement("C")
+            .withVariables(Map.of("fizz", "buzz"), "C")
+            .and()
+            .terminateElement(activatedUserTask.getKey())
+            .send();
+
+    // then
+    assertThatNoException()
+        .describedAs("Expect that modification command is not rejected")
+        .isThrownBy(command::join);
```
Comment (saig0) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/ModifyProcessInstanceTest.java:95]: :wrench: We could verify that the activation and termination instructions are applied by using the `RecordingExporter`.  

Currently, the test case would pass if the activate, terminate, or variable instructions are ignored (in the client, the gateway, or the broker).

**Review by saig0 — APPROVED** [review]
@korthout thank you for adding the test cases. :+1: And for fixing a bug. :tada: 

We could extend the test cases to verify that all given instructions are applied. It is not about verifying the logic exhaustively but to confirm that all command arguments are passed from the client to the broker. 

If you don't have time before your vacation, we can merge this PR first and improve the tests later. :palm_tree:

**Comment by korthout** [comment]
>And for fixing a bug.

It seems I made a mistake and accidentally continued on top of these changes:
- https://github.com/camunda/zeebe/pull/10353

>We could extend the test cases to verify that all given instructions are applied. It is not about verifying the logic exhaustively but to confirm that all command arguments are passed from the client to the broker.
>
>If you don't have time before your vacation, we can merge this PR first and improve the tests later.

So these changes have already been merged. I'll make sure to open an issue to improve those tests for CreateProcessInstance.

I'll also rebase this PR on top of `main`. Sorry about that @saig0

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/ModifyProcessInstanceTest.java (hunk 3 above)
Comment (korthout) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/ModifyProcessInstanceTest.java:99]: I've expanded the test case 👍

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceTest.java (hunk 2 above)
Comment (korthout) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CreateProcessInstanceTest.java]: I've opened #10382 for this

**Comment by korthout** [comment]
I've expanded the test case, please have another look @saig0

**Review by saig0 — APPROVED** [review]
@korthout LGTM :+1: 

bors merge

**Pull request merged**